lb_content_resolver/tag_search.py

```diff
@@ -0,0 +1,181 @@
+                    matching_recordings.append(recording)
+                else:
+                    over_recordings.append(recording)
+            else:
+                under_recordings.append(recording)
+
+        # If we have enough recordings, we're done!
+        if len(matching_recordings) >= num_recordings:
+            return plist(matching_recordings)
+
+        # We don't have enough recordings, see if we can pick the ones outside
+        # of our desired range in a best effort to make a playlist.
+        # Keep adding the best matches until we (hopefully) get our desired number of recordings
+        while len(matching_recordings) < num_recordings:
+            if under_recordings:
+                under_diff = begin_percent - under_recordings[-1]["percent"]
+            else:
+                under_diff = 1.0
+
+            if over_recordings:
+                over_diff = over_recordings[-1]["percent"] - end_percent
+            else:
+                over_diff = 1.0
+
+            if over_diff == 1.0 and under_diff == 1.0:
+                break
+
+            if under_diff < over_diff:
+                matching_recordings.insert(0, under_recordings.pop(-1))
+            else:
+                matching_recordings.insert(len(matching_recordings), over_recordings.pop(0))
+
+        return plist(matching_recordings)
+
+    def or_search(self, tags, min_popularity=None, max_popularity=None):
+        """
+            Return the sql query that finds recordings using the OR operator
+        """
+
+        query = """WITH recording_ids AS (
+                        SELECT DISTINCT(recording_id)
+                          FROM tag
+                          JOIN recording_tag
+                            ON recording_tag.tag_id = tag.id
+                          JOIN recording
+                            ON recording.id = recording_tag.recording_id
+                         WHERE name in (%s)
+                   )
+                       SELECT recording_mbid
+                            , popularity AS percent
+                            , subsonic_id
+                            , recording_name
+                            , artist_name
+                         FROM recording
+                         JOIN recording_ids
+                           ON recording.id = recording_ids.recording_id
+                         JOIN recording_metadata
+                           ON recording.id = recording_metadata.recording_id
+                         JOIN recording_subsonic
+                           ON recording.id = recording_subsonic.recording_id
+                           %s
+                     ORDER BY popularity DESC"""
+
+        if min_popularity is not None and max_popularity is not None:
+            pop_clause = """WHERE popularity >= %.4f AND popularity < %.4f""" % \
+                (min_popularity, max_popularity)
+        else:
+            pop_clause = ""
+
+        return query, [*tags], pop_clause
```

Comment:
[*tags] looks redundant

Reply:
That is correct, because I want them in a list, not exploded.

Reply:
why not just use tags then?

Reply:
Aawhabwbabababababhahahahahaham, because when I see *<variable> I immediately think in C, not python. lol. Fixed now.